tracer/test/Datadog.Trace.Tests/Logging/StackTraceRedactorTests.cs

```diff
@@ -0,0 +1,130 @@
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using Datadog.Trace.Logging.Internal.Sinks;
+using FluentAssertions;
+using Xunit;
+
+namespace Datadog.Trace.Tests.Logging;
+
+public class StackTraceRedactorTests
+{
+    [Fact]
+    public void Redact_IsEmptyForEmptyException()
+    {
+        var ex = new Exception();
+        var stackTrace = new StackTrace(ex);
+
+        var sb = new StringBuilder();
+        StackTraceRedactor.Redact(sb, stackTrace);
+        var redacted = sb.ToString();
+
+        redacted.Should().BeEmpty();
+    }
+
+    [Theory]
+    [MemberData(nameof(TestData.ToStringTestData), MemberType = typeof(TestData))]
+    public void Redact_ContainsExpectedStrings(StackTrace stackTrace, string expectedToString)
+    {
+        var sb = new StringBuilder();
+        StackTraceRedactor.Redact(sb, stackTrace);
+        var redacted = sb.ToString();
+
+        if (expectedToString.Length == 0)
+        {
+            redacted.Should().BeEmpty();
+            return;
+        }
+
+        redacted.Should().Contain(expectedToString);
+        redacted.Should().EndWith(Environment.NewLine);
+        var redactedFrames = redacted.Split(new[] { Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries);
+        var allowedPrefixes = new[] { "Datadog", "Microsoft", "System", "REDACTED" };
+        redactedFrames.Should().OnlyContain(x => allowedPrefixes.Any(x.Contains));
+
+        var originalFrames = stackTrace.ToString().Split(new[] { Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries);
+        redactedFrames.Length.Should().Be(originalFrames.Length);
+        for (var i = 0; i < redactedFrames.Length; i++)
+        {
+            if (redactedFrames[i].Contains("REDACTED"))
+            {
+                continue;
+            }
+
+            redactedFrames[i].Should().Be(originalFrames[i]);
+        }
+    }
+
+    [Fact]
+    public unsafe void FunctionPointerSignature_ContainsExpectedString()
+    {
+        // This is separate from Redact_ContainsExpectedStrings since unsafe cannot be used for iterators
+        var stackTrace = TestData.FunctionPointerParameter(null);
+        Assert.Contains("Datadog.Trace.Tests.Logging.StackTraceRedactorTests.TestData.FunctionPointerParameter(IntPtr x)", stackTrace.ToString());
+    }
+
+    [MethodImpl(MethodImplOptions.NoOptimization | MethodImplOptions.NoInlining)]
+    private static Exception InvokeException()
+    {
+        try
+        {
+            ThrowException();
+            return null;
+        }
+        catch (Exception ex)
+        {
+            return ex;
+        }
+    }
+
+    private static void ThrowException() => throw new Exception();
+
+    public static class TestData
+    {
+        public static IEnumerable<object[]> ToStringTestData()
+        {
+            yield return new object[] { new StackTrace(InvokeException()), "Datadog.Trace.Tests.Logging.StackTraceRedactorTests.ThrowException()" };
```

Comment:
Should we add a test case with an actual stacktrace that would contain user code as well just in case?

Reply:
> Should we add a test case with an actual stacktrace that would contain user code as well just in case?

The stack traces generated here do include "user" code, in that they include XUnit in the stack trace, which gets redacted.

However I added testing of explicit methods/properties here https://github.com/DataDog/dd-trace-dotnet/pull/3751/commits/c4f9a682a37c25d082d002965788ca9403a469e2 so we can be more confident it's working as expected

Reply:
> XUnit

Oh indeed, clearly didn't think about it 

> However I added testing of explicit methods/properties here

🙇